### ISSUE TITLE
Tasks & Context integration

### ISSUE THREAD

**Issue by JG127** [microsoft/azure-devops-intellij#260]
It would immensely help development if this plugin would integrate Azure work items with contexts, as it is already the case for other similar online services.

**Comment by ForNeVeR** [comment]
Could you please explain, which _contexts_ are you talking about? Is it about [these ones](https://docs.microsoft.com/en-us/powershell/azure/context-persistence?view=azps-3.1.0)? How would you like the plugin to integrate work items with them?

**Comment by JG127** [comment]
My bad, I should have been more specific. I meant the tasks and contexts of Jetbrains IDE (Intellij, PyCharm, ...). They put this under Tools > Tasks & Contexts. 

The IDE aims to be a one stop shop for developers who are using ticket/issue/... systems. So they won't have to log in into the system to update the tickets. But rather do everything from the IDE.

This typically starts with adding a server (Tools > Tasks & Contexts > Configure Servers > plus sign button). The IDE will propose a list of supported servers; Github, Jira, GitLab, etc, ... What this issue is about is to add Azure Devops in there.

The IDE keeps track of the task (e.g. Work Item for Azure) and the IDE context of the task. Such that resuming a task resets the IDE as it was the last time for that task. 

A task can be provided with additional comments, be closed, ...

**Comment by baumblatt** [comment]
Hello,

This will be an awesome feature for this plugin, better if associated with #176 for working with ssh as well.

Best regards,
Bernardo Baumblatt.

**Comment by lordlycastle** [comment]
@ForNeVeR Are you the official Microsoft rep responsible for this plugin? If you work on this task, the plugin would not need to be maintained any more.

**Comment by ForNeVeR** [comment]
@lordlycastle, sorry, but no, I'm not affiliated with Microsoft. So, everything I know about the status of this task is reflected by its status in the issue tracker. There's no assignee here, and currently I'm not aware of any plans to implement this feature.

Moreover, I can't see how the plugin is "not need to be maintained" if you ask for additional features to be added into the plugin.

**Comment by lordlycastle** [comment]
@ForNeVeR  It's not needed because Jetbrains has added native support for tasks. And this plugin is far worse and poorly integrated so it would be better to use their native tool. All that needs to be done is setup all the endpoints. AzureDevops has those endpoints I checked (didn't check all; just saw the REST API docs).

**Comment by JG127** [comment]
With "native support" is meant that soon we're about to see "Azure Devops" in the list of supported Task servers ?

**Comment by lordlycastle** [comment]
@JG127  Yes. All the API exists. It's a matter of configuration. I tested it with MVP, and it works because all of it is just REST API. Someone from MS needs to realise and ask their bosses to invest some effort in it. People like you and I won't have the time or inner know-how to make it perfect.

**Comment by lordlycastle** [comment]
[All the API](https://docs.microsoft.com/en-us/rest/api/azure/devops/work/taskboard%20work%20items/list?view=azure-devops-rest-6.0) can be found and tested [through Postman](https://github.com/hkamel/azuredevops-postman-collections)(Currently 5.0; needs update for 6.0).

In 5.0 there's no way to get list of work items without knowing ID first and only way to get those is through Backlog. No way to get current taskboard/iteration (sprint) work item ID.

In 6.0 you can get list of taskboard/iteration work items. But first you need the iteration ID which is another request so no way to get work items in single request; which is what IntelliJ needs. And taskboard request is in preview and I can't get it work for my Team so it's listing all Team's tasks in that iteration. We also need a way to get only new/active tasks; I can't find a way to filter.

---
Best configuration for IntelliJ would to have one request that lists all work items in current sprint with all related information without pagenation. But getting a list of IDs then information for each work items also works and I can set it up.

The problem are: 

 1. for Azure currently you need to make 2 requests to get work item IDs (1. get current iterations. 2. get work items on task board for that iteration). So you have to manually set iteration.
 2. We also need boolean value for Closes status for IntelliJ. They have `fields."System.State": "Closed"` (note dot is in name 🤨 ). I have no idea how to convert that to true/false value. Best I got was `$.fields.[?(@.['System.State']=='Closed')]` but that gives an error.
 3. `['System.Description']` field are optional; even if you specify it manually. It'll return it some not for others, so you have to remove it complete as IntelliJ will log an error and stop fetching tasks.


So that is why it doesn't exist as a server yet.... Thanks Microsoft 🙄 

![SS 2020-09-22T22 32 05@2x](https://user-images.githubusercontent.com/11729541/93934584-ab6dbd80-fd23-11ea-9c99-3a60f25654f1.png)

**Comment by JG127** [comment]
Same old egocentric MS attitude. It's all over Azure too. That's such a pity.

**Comment by lordlycastle** [comment]
Good news guys! I created a middleware layer that adds support for tasks for Azure DevOps. 

IntelliJ queries this middle layer and this middle layer does all the requests to Azure and combines all the data so IntelliJ can get all current sprint's tasks in one request. And I'm happy to say it is totally worth it for switching b/w tasks!

Currently I've deployed it to Google Cloud Functions as it's simple and free (as in it's so small it'll be in free tier). 

If people need this can be modified such that you have a local server running this middle layer and IntelliJ can call that. 

[It's very rough but it works so I'm content.](https://github.com/lordlycastle/azure-intellij-tasks)

**Comment by CaptaiNiveau** [comment]
Is there any news on this?
